OpenRA.Mods.Common/Activities/FindResources.cs

```diff
@@ -52,40 +52,59 @@ public override Activity Tick(Actor self)
 			var searchRadius = harv.LastOrderLocation.HasValue ? harvInfo.SearchFromOrderRadius : harvInfo.SearchFromProcRadius;
 			var searchRadiusSquared = searchRadius * searchRadius;
 
-			// Find harvestable resources nearby:
-			var path = self.World.WorldActor.Trait<IPathFinder>().FindPath(
-				PathSearch.Search(self.World, mobileInfo, self, true)
-					.WithHeuristic(loc =>
+			// Pick the available harvestable positions inside the harvestable radius from the selected point.
+			var harvestablePositions = self.World.Map.FindTilesInCircle(searchFromLoc, searchRadius)
+				.Where(loc =>
+				{
+					var resType = resLayer.GetResource(loc);
+					if (resType == null)
+						return false;
+
+					// Can the harvester collect this kind of resource?
+					if (!harvInfo.Resources.Contains(resType.Info.Name))
+						return false;
+
+					if (territory != null)
 					{
-						// Avoid this cell:
-						if (avoidCell.HasValue && loc == avoidCell.Value)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
+						// Another harvester has claimed this resource:
+						ResourceClaim claim;
+						if (territory.IsClaimedByAnyoneElse(self, loc, out claim))
+							return false;
+					}
 
-						// Don't harvest out of range:
-						var distSquared = (loc - searchFromLoc).LengthSquared;
-						if (distSquared > searchRadiusSquared)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost * 2;
+					return true;
+				}).ToList();
 
-						// Get the resource at this location:
-						var resType = resLayer.GetResource(loc);
-						if (resType == null)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
+			List<CPos> path;
 
-						// Can the harvester collect this kind of resource?
-						if (!harvInfo.Resources.Contains(resType.Info.Name))
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
+			// If no harvestable position could be found, then return an empty path
+			if (harvestablePositions.Count == 0)
+				path = new List<CPos>();
+			else
+			{
+				var proc = harv.ClosestProc(self, null);
+
+				// Pick the closest harvestable position from the actor
+				var closestHarvestablePosition = harvestablePositions.MinBy(loc =>
+					proc == null ? (loc - self.Location).Length : (loc - proc.Location).Length);
```

Comment:
Obvious if minor perf:
- Lift the conditional ```perf == null``` outside the enumeration.
- Use LengthSquared.